rust/core-lib/src/plugins.rs

```diff
@@ -0,0 +1,388 @@
+    let mut path_base = env::current_exe().unwrap();
+    path_base.pop();
+    let make_path = |p: &str| -> PathBuf {
+        let mut pb = path_base.clone();
+        pb.push(p);
+        pb
+    };
+
+    vec![
+        PluginDescription::new("syntect", "0.0", make_path("xi-syntect-plugin")),
+        PluginDescription::new("braces", "0.0", make_path("bracket_example.py")),
+        PluginDescription::new("spellcheck", "0.0", make_path("spellcheck.py")),
+        PluginDescription::new("shouty", "0.0", make_path("shouty.py")),
+    ]
+}
+
+/// Describes attributes and capabilities of a plugin.
+///
+/// Note: - these will eventually be loaded from manifest files.
+#[derive(Debug, Clone)]
+pub struct PluginDescription {
+    pub name: String,
+    version: String,
+    //scope: PluginScope,
+    // more metadata ...
+    /// path to plugin executable
+    pub exec_path: PathBuf,
+}
+
+impl PluginDescription {
+    fn new<S, P>(name: S, version: S, exec_path: P) -> Self
+        where S: Into<String>, P: Into<PathBuf> {
+            PluginDescription {
+                name: name.into(),
+                version: version.into(),
+                exec_path: exec_path.into()
+            }
+    }
+
+    /// Starts the executable described in this `PluginDescription`.
+    fn launch<W, C>(&self, manager_ref: &Arc<Mutex<PluginManager<W>>>, buffer_id: &str, completion: C)
+    where W: Write + Send + 'static, C: FnOnce(Result<PluginRef<W>, &'static str>) + Send + 'static {
+    //TODO: a real result type
+
+        let path = self.exec_path.clone();
+        let buffer_id = buffer_id.to_owned();
+        let manager_ref = manager_ref.clone();
+        let description = self.clone();
+
+        thread::spawn(move || {
+            print_err!("starting plugin at path {:?}", path);
+            let mut child = Command::new(&path)
+                .stdin(Stdio::piped())
+                .stdout(Stdio::piped())
+                .spawn()
+                .expect("plugin failed to start");
+            let child_stdin = child.stdin.take().unwrap();
+            let child_stdout = child.stdout.take().unwrap();
+            let mut looper = RpcLoop::new(child_stdin);
+            let peer = looper.get_peer();
+            peer.send_rpc_notification("ping", &Value::Array(Vec::new()));
+            let plugin = Plugin {
+                peer: peer,
+                //TODO: I had the bright idea of keeping this reference but
+                // I'm not sure exactly what to do with it (stopping the plugin is one thing)
+                process: child,
+                manager: Arc::downgrade(&manager_ref),
+                description: description,
+                buffer_id: buffer_id,
+            };
+            let mut plugin_ref = PluginRef(Arc::new(Mutex::new(plugin)));
+            completion(Ok(plugin_ref.clone()));
+            looper.mainloop(|| BufReader::new(child_stdout), &mut plugin_ref);
+        });
+    }
+}
+
+
+type PluginName = String;
```

Comment:
I wonder if we should use a struct/tuple struct later. At one point it may be nicer to create a `usize` for a Plugin Id (or any other id for that matter) since they are smaller and easier to send around. If we encapsulate it into a struct, it's much easier to adjust the structure later if needed.